spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/integration/PubSubHeaderMapper.java

```diff
@@ -0,0 +1,96 @@
+/*
+ *  Copyright 2018 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.pubsub.integration;
+
+import java.util.Arrays;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import org.springframework.integration.mapping.HeaderMapper;
+import org.springframework.integration.util.PatternMatchUtils;
+import org.springframework.messaging.MessageHeaders;
+import org.springframework.util.Assert;
+
+public class PubSubHeaderMapper implements HeaderMapper<Map<String, String>> {
+
+	/**
+	 * Header patterns to map in {@link #fromHeaders(MessageHeaders, Map)}.
+	 */
+	private String[] outboundHeaderPatternsToMap = {"*",
+			"!" + MessageHeaders.ID,
+			"!" + MessageHeaders.TIMESTAMP};
```

Comment:
Are you sure that the order must be like this?
Do you have a test-case to confirm that we don't send `id` and `timestamp` by default?

See `smartMatch` JavaDocs:
```
 * Pattern match against the supplied patterns; also supports negated ('!')
 * patterns. First match wins (positive or negative).
```

Therefore that asterisk in the first position will always match to `id` and `timestamp` and these negative patterns will never be accessed.

See SI-AWS solution one more time:
https://github.com/spring-projects/spring-integration-aws/blob/master/src/main/java/org/springframework/integration/aws/support/AbstractMessageAttributesHeaderMapper.java#L51